src/components/TodoItem.tsx

```diff
@@ -0,0 +1,154 @@
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import React, {
+  useContext, useState, useRef, useEffect,
+} from 'react';
+import { Todo } from '../types/Todo';
+import { TodosContext } from '../contexts/TodosContext';
+
+type Props = {
+  todo: Todo;
+};
+
+enum TodoStatus {
+  View = 'view',
+  Editing = 'editing',
+  Completed = 'completed',
+}
+
+enum ToggleStatus {
+  View = 'toggle-view',
+  Editing = 'toggle-editing',
+  Completed = 'toggle-completed',
+}
+
+type InitialStatus = {
+  input: TodoStatus,
+  toggle: ToggleStatus,
+};
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const initialStatus = {
+    input: todo.completed ? TodoStatus.Completed : TodoStatus.View,
+    toggle: todo.completed ? ToggleStatus.Completed : ToggleStatus.View,
+  };
+
+  const { todos, setTodos } = useContext(TodosContext);
+  const [todoStatus, setTodoStatus] = useState<InitialStatus>(initialStatus);
+  const [newTitle, setNewTitle] = useState(todo.title);
+
+  const editInputRef = useRef<HTMLInputElement>(null);
+
+  const { id, title } = todo;
+
+  const handleTodoDelete = (todoId: number) => {
+    const filteredTodos
+      = todos.filter(currentTodo => currentTodo.id !== todoId);
+
+    setTodos(filteredTodos);
+  };
+
+  const handleToggle = () => {
+    const modifiedTodos = todos.map(currentTodo => {
+      if (currentTodo.id === todo.id) {
+        return {
+          ...currentTodo,
+          completed: !currentTodo.completed,
+        };
+      }
+
+      return currentTodo;
```

Comment:
can be more simplify

```suggestion
      currentTodo.id === todo.id ?
          {
          ...currentTodo,
          completed: !currentTodo.completed,
        } : currentTodo
```